distributed/shuffle/multi_file.py

```diff
@@ -0,0 +1,282 @@
+
+        We do this until we're done.  This coroutine runs in the background.
+
+        See Also
+        --------
+        process: does the actual writing
+        """
+        with log_errors():
+
+            while not self._done:
+                with self.time("idle"):
+                    if not self.shards:
+                        await asyncio.sleep(0.1)
+                        continue
+
+                    await self.queue.get()
+
+                id = max(self.sizes, key=self.sizes.get)
+                shards = self.shards.pop(id)
+                size = self.sizes.pop(id)
+
+                future = asyncio.ensure_future(self.process(id, shards, size))
+                del shards
+                self._futures.add(future)
+                async with self.condition:
+                    self.condition.notify()
+
+    async def process(self, id: str, shards: list, size: int):
+        """Write one buffer to file
+
+        This function was built to offload the disk IO, but since then we've
+        decided to keep this within the event loop (disk bandwidth should be
+        prioritized, and writes are typically small enough to not be a big
+        deal).
+
+        Most of the logic here is about possibly going back to a separate
+        thread, or about diagnostics.  If things don't change much in the
+        future then we should consider simplifying this considerably and
+        dropping the write into communicate above.
+        """
+
+        with log_errors():
+            # Consider boosting total_size a bit here to account for duplication
+            while id in self.active:
+                await asyncio.sleep(0.01)
+
+            self.active.add(id)
+
+            def _():
+                with open(
+                    self.directory / str(id), mode="ab", buffering=100_000_000
+                ) as f:
+                    for shard in shards:
+                        self.dump(shard, f)
+                    # os.fsync(f)  # TODO: maybe?
+
+            start = time.time()
+            with self.time("write"):
+                _()
+            #     await offload(_)
+            stop = time.time()
+
+            self.diagnostics["avg_size"] = (
+                0.98 * self.diagnostics["avg_size"] + 0.02 * size
+            )
+            self.diagnostics["avg_duration"] = 0.98 * self.diagnostics[
+                "avg_duration"
+            ] + 0.02 * (stop - start)
+
+            self.active.remove(id)
+            self.bytes_written += size
+            self.total_size -= size
+            MultiFile.total_size -= size
+            async with self.condition:
+                self.condition.notify()
+            await self.queue.put(None)
+
+    def read(self, id):
+        """Read a complete file back into memory"""
+        parts = []
+
+        try:
+            with self.time("read"):
+                with open(
+                    self.directory / str(id), mode="rb", buffering=100_000_000
```

Comment:
I don't think buffering will do much on read, unless pyarrow is doing something very special. Shouldn't harm, though.

Reply:
Can you expand on this?  My guess is that buffering with a file in read mode would mean that we read in a larger amount of data when we do any read.  I'd much rather do a single giant read (the entire file really) than do lots of small reads.